### DISCUSSION TITLE
Algorithm implementation in Minecraft

### DISCUSSION THREAD

**Discussion by RCoder01** [PlummersSoftwareLLC/Primes#627] [Q&A] [answered]
I was able to code an implementation of the sieve of Eratosthenes in Minecraft commands, sort of as a joke. It does genuinely work though, and I believe it also follows the rules of this project. I was wondering if this would be allowed to be uploaded to the project given the nature of how unconventional it is compared to a 'normal' programming language, even many esoteric languages. The compiler/interpreter is quite large and resource intensive because it's a video game not meant only to be a fairly basic programming language.

All the files required to duplicate the program would only be about 15 KB, so it's not like you would have to upload the entirety of Minecraft to the project, but a copy of the game would be required to run the program. I'm not too familiar with docker, but I would be surprised if it was possible to configure it to run a Minecraft command program, so if that's a deal breaker, that's fair. (And if it is possible, I will certainly need help making it)

I have genuinely tried to make my implementation as fast as I can (without essentially hardcoding values), and it takes a bit over 3 mins to run one iteration up to 1,000,000 on my computer (although it probably could be brought down below 3 mins with some configuration testing; also takes 13 seconds for up to 100,000 and ~.8 seconds for 10,000), so obviously it isn't practical and won't be winning any speed awards (if that wasn't already clear), but I still think it would be an interesting addition to this project.

**Comment by rbergen** [comment, marked as answer]
Generally speaking, we don't add solutions that can't be included in the automated benchmark reporting, certainly _if alternatives are available that can be_. Examples of this are:

- A solution in Lua Roblox (#50), which was later converted to a "plain Lua" solution (Lua Roblox is commercially licensed and doesn't run in Docker, plain Lua is freeware and works in Docker)
- A solution that conceptually started as a MATLAB solution (#72), that ended up being merged as an Octave solution (MATLAB is commercially licensed, Octave is freeware)
- A version of a COBOL solution targeting IBM mainframes (#333), that ended up being a show and tell post (we had a COBOL solution that runs in the benchmark context but no hardware running z/OS)
- A solution in Oracle PL/SQL (#462), which we didn't merge (Oracle is commercially licensed, and other freeware SQL dialects were already included)
- A solution in APL (#549) that requires Dyalog to run; the PR for it was closed by the author (I'd still love to add the language). Dyalog is commercially licensed and a freeware APL implementation (GNU APL) is available, but the solution would require adaptations to work on it. 

As you can tell from these examples, the most common reason for not including solution is not technical, but related to licensing. I would think that this would actually also be the main hurdle concerning Minecraft.

All that said, as far as I know there is no freeware alternative available for Minecraft and we do have one solution in place without a Dockerfile (written in Delphi). The best thing to do may be to put a PR together, so we can see if and in what form we can add your Minecraft implementation to the project.
There is one important thing that will apply, which is that we don't merge binary files. This means that the solution would need to consist of the text-based source code and whatever instructions need to be followed to get that to run.

**Reply by RCoder01** [reply]
I'm pretty confident there are no legal freeware alternatives of Minecraft, but looking into it further, it should be possible to run my program on just the server version of the game. Although I'm nowhere near an expert in licensing, it seems to me that downloading and using the server JAR [from this page](https://www.minecraft.net/en-us/download/server) only requires agreeing to the [EULA](https://account.mojang.com/documents/minecraft_eula) and [privacy policy](https://privacy.microsoft.com/en-us/privacystatement), which allows its non-commerical use (which I think this project falls under). 

With the server version, I think it would be possible to make a shell or python script to automate running the server and the program and run that from the Dockerfile, but I have no idea how to go about doing that.
Also, the only files consisting my program are .json, .mcmeta, and .mcfunction, all of which are text files so that shouldn't be an issue. The Minecraft server JAR itself is binary, but that should be downloaded from Minecraft directly per the EULA, not distributed through this project.

**Reply by rbergen** [reply]
I look forward to seeing the PR for this. I suggest we start with the source code as you have it, and just take it from there.

**Comment by minecraftchest1** [comment]
May you share a link to the repro with the implementation? I am interested in seeing how it works. I may also be able to get it somewhat automated.

**Reply by RCoder01** [reply]
I have [a fork of the repo on my profile](https://github.com/RCoder01/Primes/tree/drag-race/PrimeMinecraftDatapack/solution_1) with the program in the PrimeMinecraftDatapack/solution_1 folder accompanied by a readme with instructions on how to use it. I've been busy with college applications since just after I originally opened this thread, so I haven't had the time to work on automating it, and help would be greatly appreciated.

**Reply by rbergen** [reply]
@RCoder01 What would you like to automate?

**Reply by RCoder01** [reply]
@rbergen I'm not sure if "automate" is the right word, but I mean to set up a dockerfile and/or a script to automatically set up and run the code because it might be confusing to someone unfamiliar with Minecraft.

**Reply by rbergen** [reply]
@RCoder01 I think "automate" is the right word. :) I could also help out with putting a Dockerfile together, even though we will have to disable its inclusion in the automated benchmark runs due to licensing concerns we've discussed before.
Is this indeed something you'd like me to work on?

**Reply by RCoder01** [reply]
@rbergen I would certainly appreciate that.

**Reply by rbergen** [reply]
I'm getting stuck with getting Minecraft server to execute the function when stdin is redirected, which I need to do to be able to feed the Minecraft server console commands that need to be issued. 

When stdin is redirected, the following is shown and logged after issuing `function sieve:full`:
```
[22:37:21] [Server thread/INFO]: [Server] started
[22:37:22] [Server thread/WARN]: Exception loading entity:
z: Loading entity NBT
        at atg.g(SourceFile:1776) ~[server.jar:?]
        at atk.a(SourceFile:485) ~[server.jar:?]
        at ad.a(SourceFile:448) ~[server.jar:?]
        at atk.a(SourceFile:484) ~[server.jar:?]
        at atk.b(SourceFile:577) ~[server.jar:?]
        at atk.a(SourceFile:528) ~[server.jar:?]
        at zy.a(SourceFile:65) ~[server.jar:?]
        at zy.a(SourceFile:48) ~[server.jar:?]
        at com.mojang.brigadier.CommandDispatcher.execute(CommandDispatcher.java:263) ~[server.jar:?]
        at di$b.a(SourceFile:96) ~[server.jar:?]
        at di$b.a(SourceFile:90) ~[server.jar:?]
        at xi$b.a(SourceFile:137) ~[server.jar:?]
        at xi$a.b(SourceFile:187) ~[server.jar:?]
        at xi.a(SourceFile:87) ~[server.jar:?]
        at xi.a(SourceFile:71) ~[server.jar:?]
        at ym.a(SourceFile:42) ~[server.jar:?]
        at ym.a(SourceFile:33) ~[server.jar:?]
        at com.mojang.brigadier.CommandDispatcher.execute(CommandDispatcher.java:263) ~[server.jar:?]
        at com.mojang.brigadier.CommandDispatcher.execute(CommandDispatcher.java:177) ~[server.jar:?]
        at dm.a(SourceFile:251) ~[server.jar:?]
        at aas.bf(SourceFile:363) ~[server.jar:?]
        at aas.b(SourceFile:336) ~[server.jar:?]
        at net.minecraft.server.MinecraftServer.a(SourceFile:831) ~[server.jar:?]
        at net.minecraft.server.MinecraftServer.x(SourceFile:697) ~[server.jar:?]
        at net.minecraft.server.MinecraftServer.a(SourceFile:270) ~[server.jar:?]
        at java.lang.Thread.run(Thread.java:833) [?:?]
Caused by: java.lang.NullPointerException: Cannot invoke "bwq.j_()" because "this.t" is null
        at atg.bO(SourceFile:2222) ~[server.jar:?]
        at atg.i(SourceFile:2218) ~[server.jar:?]
        at atg.g(SourceFile:1754) ~[server.jar:?]
        ... 25 more
[22:37:22] [Server thread/INFO]: [Server] ended
[22:37:22] [Server thread/INFO]: Executed 22 commands from function 'sieve:full'
```

I've actually even written a program in C that waits with issuing commands until the Minecraft server console is ready to accept them, but that makes no difference. 

I'm afraid I'm not sufficiently knowledgeable about Minecraft development to be able to debug this.

**Reply by rbergen** [reply]
@RCoder01 Ok, it seems I've been able to make some progress. I've put some scripts and support tooling together that allow for a scripted preparation ("build") and execution ("run") on another one of my systems. I have an idea what the problem on the other system might be, but I will test soon to confirm. 

If I can get the scripts to work consistently on a few systems, the next step will be trying to put a Dockerfile together to get the solution running that way.

EDIT: I can confirm that the scripts now work on the system that consistently failed, before. It turns out that when `[Server] loaded` is reported, the Minecraft server needs a few seconds to settle down before `function sieve:full` is issued.

The current state of affairs (whatever that is when you read this :) ) can be found in [github.com/rbergen/Primes:minecraft](https://github.com/rbergen/Primes/tree/minecraft/PrimeMinecraftDatapack/solution_1).

I'll post here again when there is news to report.

**Reply by RCoder01** [reply]
@rbergen The `[Server] loaded` message is only sent when my code is loaded, so my guess is that that happens somewhere before the end of the overall server loading process, but honestly I also don't know enough about Minecraft to say that for sure. Anyways, glad you were able to figure out the issue, and thanks again for helping with this.

**Reply by rbergen** [reply]
@RCoder01 I just pushed an update to my minecraft branch that includes a working Dockerfile. 

Take a look when you have a chance, and let me know if you're happy to move forward with this. We can then discuss how to push it to the main repo.

**Reply by rbergen** [reply]
@RCoder01 I'd like to add the Minecraft solution to the project before Dave puts out his next drag race episode. Are you ok with me opening a PR on your branch to add my automation? You can open a PR on the main project after.

**Reply by RCoder01** [reply]
@rbergen Sure, that would be fine.

**Reply by rbergen** [reply]
Just opened RCoder01#1.

Note that when opening the PR on the main repo, you'll most likely have to sync your fork with the current state of it. This will be needed to get CI to succeed.

**Reply by RCoder01** [reply]
@rbergen I just opened #756.

**Reply by rbergen** [reply]
@RCoder01 Yes, reviewed and approved. I've asked @marghidanu to perform the second review, after which we can merge.

**Reply by rbergen** [reply]
Merged! 

@RCoder01 Thank you for putting this together!

**Reply by RCoder01** [reply]
Thank you @rbergen for all the automation help!
